[ZK filter] emit fast/slow response metrics for error budget SLI calculation

Winbobob opened the pull request:
Commit Message: [ZK filter] emit fast/slow response metrics for error budget SLI calculation
Additional Description: Emit fast/slow response counter metrics for each opcode used for error budget SLI calculation.
Risk Level: low
Testing: unit tests
Docs Changes: This diff will update these two docs [ZooKeeper proxy introduction doc](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/zookeeper_proxy_filter#zookeeper-proxy) and [ZooKeeper proxy proto doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto).
Release Notes: [ZK filter] emit fast/slow response metrics for error budget SLI calculation
Platform Specific Features: N/A
[API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md): Did not find a good way to make repeated message unique (`LatencyThreshold` in this case) with proto validation. Also found that enum cannot be the key of the proto map.

Winbobob commented on docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst:
```diff
@@ -76,64 +79,119 @@ following counters are available:
 -----------------------------
 
-The filter will gather latency statistics in the *<stat_prefix>.zookeeper.<opcode>_response_latency* namespace.
+The filter will gather latency statistics in the *<stat_prefix>.zookeeper.<opcode>_resp_latency* namespace.
```
Change `response_latency` -> `resp_latency` to align with the code.

Winbobob commented on docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst:
```diff
@@ -47,27 +47,30 @@ following counters are available:
   getdata_rq, Counter, Number of getdata requests
   create_rq, Counter, Number of create requests
   create2_rq, Counter, Number of create2 requests
+  createcontainer_rq, Counter, Number of createcontainer requests
```
Add missing metric names and make metric orders consistent among `*_rq`, `*_resp`, `*_resp_fast`, `*_resp_slow`

Winbobob commented on source/extensions/filters/network/zookeeper_proxy/decoder.cc:
```diff
@@ -631,7 +631,7 @@ void DecoderImpl::parseConnectResponse(Buffer::Instance& data, uint64_t& offset,
   const bool readonly = maybeReadBool(data, offset);
 
-  callbacks_.onConnectResponse(0, timeout, readonly, latency);
+  callbacks_.onConnectResponse(OpCodes::Connect, 0, timeout, readonly, latency);
```
To make `onConnectResponse` and `onResponse` able to reuse code.

Winbobob commented on source/extensions/filters/network/zookeeper_proxy/filter.cc:
```diff
@@ -18,12 +20,16 @@ namespace Extensions {
+      latency_threshold_map_(parseLatencyThresholds(latency_thresholds)),
       stat_name_set_(scope.symbolTable().makeSet("Zookeeper")),
       stat_prefix_(stat_name_set_->add(stat_prefix)), auth_(stat_name_set_->add("auth")),
-      connect_latency_(stat_name_set_->add("connect_response_latency")),
```
To make all opcode has the same latency metric name format `[opcode]_resp_latency`

Winbobob commented on api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:
```diff
@@ -39,4 +39,50 @@ message ZooKeeperProxy {
+  // If the default threshold is not set, it will be set as 100 millisecond.
+  // If the threshold of one opcode is not set, it will fallback to use the default
+  // threshold.
+  repeated LatencyThreshold latency_thresholds = 4;
```
Did not find a good way to make sure each `LatencyThreshold` is unique.

Winbobob commented:
cc @rgs1 and @JuniorHsu

JuniorHsu commented on api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:
```diff
@@ -39,4 +39,51 @@ message ZooKeeperProxy {
+
+  Opcode opcode = 1 [(validate.rules).enum = {defined_only: true}];
+
+  google.protobuf.UInt32Value threshold = 2 [(validate.rules).uint32 = {gt: 0}];
```
I think we can use `unit32` instead of protobuf?

JuniorHsu commented on api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:
```diff
@@ -39,4 +39,51 @@ message ZooKeeperProxy {
+  // If no latency threshold is set, the fast/slow response counters will always be 0.
+  // If the default threshold is not set, it will be set as 100 millisecond.
+  // If the threshold of one opcode is not set, it will fallback to use the default
+  // threshold.
```
these two fallback should be unit tested

Winbobob replied:
https://github.com/envoyproxy/envoy/pull/26594/files#diff-a0f6243d3325c5bb6bbda5bb0e9c6a47723f25d95934e2edf05a98dc978f8f4fR34-R36

JuniorHsu commented on source/extensions/filters/network/zookeeper_proxy/config.cc:
```diff
@@ -27,9 +27,12 @@ Network::FilterFactoryCb ZooKeeperConfigFactory::createFilterFactoryFromProtoTyp
   const uint32_t max_packet_bytes =
       PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config, max_packet_bytes, 1024 * 1024);
+  const Protobuf::RepeatedPtrField<
+      envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold>&
```
put the alias in filter.h
```using LatencyThreshold = envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold;
using LatencyThresholdList = Protobuf::RepeatedPtrField<LatencyThreshold>```


and replace the long type here and there

JuniorHsu commented on source/extensions/filters/network/zookeeper_proxy/filter.cc:
```diff
@@ -288,24 +325,29 @@ void ZooKeeperFilter::onCloseRequest() {
+                      {"error", std::to_string(error)}});
+}
+
+std::string ZooKeeperFilter::onResponseHelper(const OpCodes opcode,
```
it's vague to me what `onResponseHelper` does and returns.
this might be clear but i'll let you judge
```
onResponse {
opinfo = getOpinfo(opcode);
std::string = opname{};
if (opinfo != unknown) {
  opname = opinfo.name_;
  recordErrorBudget();
}
setDynamicData(...)
}
```

JuniorHsu commented on source/extensions/filters/network/zookeeper_proxy/filter.cc:
```diff
@@ -314,17 +356,39 @@ void ZooKeeperFilter::onResponse(const OpCodes opcode, const int32_t xid, const
+      auto it = config_->latency_threshold_map_.find(-1);
+      if (it != config_->latency_threshold_map_.end()) {
+        default_latency_threshold = it->second;
+      }
```
let's make `default_latency_threshold` a data member, and only calculate once when you init the filter config

JuniorHsu reviewed (changes requested):
great patch. left some nits and test enrichment.

JuniorHsu commented on source/extensions/filters/network/zookeeper_proxy/filter.h:
```diff
@@ -115,19 +177,121 @@ class ZooKeeperFilterConfig {
   const uint32_t max_packet_bytes_;
   ZooKeeperProxyStats stats_;
+  // Key: opcode enum value, value: latency threshold in millisecond.
+  absl::flat_hash_map<int32_t, uint32_t> latency_threshold_map_;
```
`const absl::flat_hash_map<int32_t, uint32_t> latency_threshold_map_;`

adisuissa commented on api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:
```diff
@@ -39,4 +39,50 @@ message ZooKeeperProxy {
+    SetWatches2 = 26;
+  }
+
+  Opcode opcode = 1 [(validate.rules).enum = {defined_only: true}];
```
Please add a short comment describing the field.

adisuissa commented on api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:
```diff
@@ -39,4 +39,50 @@ message ZooKeeperProxy {
+
+  Opcode opcode = 1 [(validate.rules).enum = {defined_only: true}];
+
+  uint32 threshold = 2 [(validate.rules).uint32 = {gt: 0}];
```
It is unclear what the threshold units are.
Can you add a short comment describing the field?

adisuissa commented on api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:
```diff
@@ -39,4 +39,50 @@ message ZooKeeperProxy {
   // if that is set. If it isn't, ZooKeeper's default is also 1Mb.
   google.protobuf.UInt32Value max_packet_bytes = 3;
+
+  // List of latency thresholds for opcodes to decide the fast/slow responses and emit metrics (used for error budget calculation).
```
Can multiple entries have the same Opcode?
If not, please clarify the comment, and add a test that validates that the config is rejected.

Winbobob replied:
I've added a valid unit test with duplicated opcode entries.

adisuissa replied:
FWIW, PGV (protoc-gen-validate) cannot check this, but the code can, and decide to reject (throw an exception) in this case. As an example: [route validity](https://github.com/envoyproxy/envoy/blob/18f89086ef5341b7d152eb3f1f5b16e2ea3749eb/source/common/router/config_impl.cc#L156-L160) is one of the cases where a config is rejected, and cannot be validated by PGV.

Overriding is a valid approach, with the downside of not detecting configuration smells until it becomes late.

Whatever approach you choose, please update the comments to describe this.

Winbobob replied:
Thanks, I prefer check the duplicated opcode in the code. Will make the code/test/comment changes accordingly.

adisuissa commented on docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst:
```diff
@@ -76,64 +79,119 @@ following counters are available:
-  removewatches_response_latency, Histogram, Opcode execution time in milliseconds
-  check_response_latency, Histogram, Opcode execution time in milliseconds
+  getdata_resp_latency, Histogram, Opcode execution time in milliseconds
+  create_resp_latency, Histogram, Opcode execution time in milliseconds
```
Changing the names of the stat names may cause a breaking change if some 3rd party is dependent on them (throughout this document and PR).
@htuch I'm not sure if stat renaming is typically considered an API breaking change or not, but are you aware of prior cases where stat names were renamed/deprecated?

htuch replied:
There is prior art. I think we don't apply API breaking change policy, but we do put this in the class of "don't break operators gratuitously", so for a simple rename for consistency we might avoid that, and anything more major might roll out under runtime protection or something like that. @mattklein123 probably remembers better than I do the case history and philosophy.

Winbobob replied:
Thank you for the clarification! I've reverted all changes related to the existing metric names. For all changes in the .rst file, I explained in my previous comment above.

adisuissa reviewed (commented):
Thanks.
Left a few API comments.

Winbobob commented:
> Thanks.
> 
> Left a few API comments.

Hi @adisuissa, please review again if you have a chance, thank you!

adisuissa commented on api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:
```diff
@@ -39,4 +39,55 @@ message ZooKeeperProxy {
+  Opcode opcode = 1 [(validate.rules).enum = {defined_only: true}];
+
+  // The latency threshold of each opcode in millisecond.
+  uint32 threshold = 2 [(validate.rules).uint32 = {gt: 0}];
```
Is there a reason to use the `uint32` type over `google.protobuf.Duration`?

Winbobob replied:
Thanks, let me make the change accordingly.

adisuissa commented on docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst:
```diff
@@ -47,27 +47,30 @@ following counters are available:
   getchildren_rq, Counter, Number of getchildren requests
   getchildren2_rq, Counter, Number of getchildren2 requests
-  remove_rq, Counter, Number of delete requests
+  delete_rq, Counter, Number of delete requests
```
Consider using the same name `remove_rq` to avoid breaking it. Same goes for `remove_resp` -> `delete_resp`

Winbobob replied:
Hi @adisuissa, actually this is not a metric name change, instead, it is a document fix. Because there is no `remove_rq` or `remove_resp` specified in the code, in contrast, the ZK filter use [delete_rq](https://github.com/envoyproxy/envoy/blob/4e9c63b738d847d066166a48ca75f5669eda0da2/source/extensions/filters/network/zookeeper_proxy/filter.h#L41) and [delete_resp](https://github.com/envoyproxy/envoy/blob/4e9c63b738d847d066166a48ca75f5669eda0da2/source/extensions/filters/network/zookeeper_proxy/filter.h#L70).

If some users use the ZK filter, they will not be able to find metrics for `remove_rq` or `remove_resp`. They can only find `delete_rq` or `delete_resp`.

HTH

adisuissa reviewed (commented):
Apologies for the late response. Here's another API review round. Thanks!

Winbobob commented:
> Apologies for the late response. Here's another API review round. Thanks!

Hi @adisuissa, just addressed your comments. Please review again when you have a chance, thank you!

adisuissa commented on api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:
```diff
@@ -39,4 +40,60 @@ message ZooKeeperProxy {
+  // The latency threshold of each opcode.
+  google.protobuf.Duration threshold = 2 [(validate.rules).duration = {
+    required: true
+    gt {}
```
Can you verify if the expected resolution is in milliseconds or nanoseconds? 
If it is the former, consider having: `gte {nanos: 1000000}` as a constraint.

Winbobob replied:
Thank you for your suggestions!
I think as long as the user specify the threshold, it doesn't matter whether it is in nanosecond, millisecond or in second. This is because in the code I use `PROTOBUF_GET_MS_REQUIRED`, it should convert the user-specified time in millisecond anyway.

adisuissa replied:
Wouldn't this cast a value that is below a millisecond to 0?
It seems that this is an invalid case according to your constraints here, no?

Winbobob replied:
Yeah, make sense, thank you! Just updated the code and the test.

adisuissa reviewed (commented):
API LGTM, modulo constraint comment.
@JuniorHsu can you PTAL as code-owner?
/assign @JuniorHsu

JuniorHsu commented on source/extensions/filters/network/zookeeper_proxy/filter.h:
```diff
@@ -90,13 +146,16 @@ struct ZooKeeperProxyStats {
 };
 
+using LatencyThresholdList = const Protobuf::RepeatedPtrField<
+    envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold>&;
```
move `const` and `&` outside of alias and `const LatencyThresholdList&` elsewhere

JuniorHsu commented on source/extensions/filters/network/zookeeper_proxy/filter.cc:
```diff
@@ -32,44 +37,197 @@ ZooKeeperFilterConfig::ZooKeeperFilterConfig(const std::string& stat_prefix,
+  absl::flat_hash_map<int32_t, std::chrono::milliseconds> latency_threshold_map;
+  for (const auto& threshold : latency_thresholds) {
+    switch (threshold.opcode()) {
+    case envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold::Default:
```
let's try some DRY

introducing 
```
using LatencyThreshold_Opcode = envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold_Opcode;
using LatencyThreshhold = envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold;

int32 getOpCodeIndex(LatencyThreshold_Opcode code) {
  static constexpr absl::flat_hash_map<LatencyThreshold_Opcode, int32_t> opcode_map = {
  // Use a special key to store the default latency threshold.
    {LatencyThreshold::Default, -999}, 
   {LatencyThreshold::Connect, 0}, ...
  }
 }

  ASSERT(opcode_map.find(code) != opcode_map.end());
  return opcode_map[code];
}
``` 
and here we could do
```
latency_threshold_map[getOpcodeIndex(threshold.opcode()] =
          std::chrono::milliseconds(PROTOBUF_GET_MS_REQUIRED(threshold, threshold));
```

JuniorHsu replied:
Also could we move the check `// Check duplicated opcodes in config.` to here?

JuniorHsu replied:
i,e., remove the `std::set` section and throw exception here if dup

Winbobob replied:
> Also could we move the check `// Check duplicated opcodes in config.` to here?

I think the opcode duplication check related to config validation, so I put it in `config.cc`.

JuniorHsu commented on source/extensions/filters/network/zookeeper_proxy/filter.cc:
```diff
@@ -293,6 +465,22 @@ void ZooKeeperFilter::onConnectResponse(const int32_t proto_version, const int32
                                         const std::chrono::milliseconds& latency) {
   config_->stats_.connect_resp_.inc();
 
+  if (!config_->latency_threshold_map_.empty()) {
```
let's move those logic into ZookeeperFilterConfig by following steps
(1) move `default_latency_threshold_` to Config as it belongs to the proto
(2) introduce a enum class {NONE, SLOW, FAST}
(3) open an public function from ZookeeperFilterConfig like errorbudgetDecision and we can do
switch (config->errorbudgetDecision(OpCode::Connect, latency) {
  ...
}
(4) move `latency_threshold_map_` into private section


I know this is tedious but I though we can do better abstraction and DRY over connect and other opcode. wdyt?

Winbobob replied:
Thank you very much for your suggestions!
I think they are really good ideas, which make the code high cohesion and low coupling.

JuniorHsu reviewed (changes requested):
mainly DRY stuff

Winbobob commented on source/extensions/filters/network/zookeeper_proxy/filter.cc:
```diff
@@ -19,55 +21,170 @@ namespace NetworkFilters {
 }
 
+int32_t ZooKeeperFilterConfig::getOpCodeIndex(LatencyThreshold_Opcode opcode) {
+  static absl::flat_hash_map<LatencyThreshold_Opcode, int32_t> opcode_map = {
```
I did not use `constexpr` here because the compiler says `constexpr variable cannot have non-literal type 'const absl::flat_hash_map<LatencyThreshold_Opcode, int32_t>'`

JuniorHsu commented on source/extensions/filters/network/zookeeper_proxy/filter.h:
```diff
@@ -90,13 +146,21 @@ struct ZooKeeperProxyStats {
 
+enum class ErrorBudgetResponseType { FAST, SLOW, NONE };
+
+using LatencyThresholdList = const Protobuf::RepeatedPtrField<
```
remove const or `ConstLatencyThresholdList` but the former is preferred

Winbobob replied:
Thanks, just made the change to remove const from alias.

JuniorHsu commented on source/extensions/filters/network/zookeeper_proxy/filter.h:
```diff
@@ -90,13 +146,21 @@ struct ZooKeeperProxyStats {
+    envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold>;
+using LatencyThreshold_Opcode =
+    envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold_Opcode;
+using LatencyThreshold = envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold;
```
my bad. could we try 
`using envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold_Opcode;` to remove the redundant assignment on the above two?

Winbobob replied:
I think `LatencyThreshold_Opcode` and `LatencyThreshold` are two different namespaces.
I use below in the end:
```
using envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold;
using envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold_Opcode;
```

JuniorHsu replied:
yeah you need to declare both.

JuniorHsu commented on source/extensions/filters/network/zookeeper_proxy/filter.h:
```diff
@@ -122,12 +188,26 @@ class ZooKeeperFilterConfig {
   const Stats::StatName unknown_opcode_latency_;
 
+  ErrorBudgetResponseType errorBudgetDecision(const OpCodes opcode,
+                                              const std::chrono::milliseconds& latency);
```
we can just copy the latency. i.e., no '&'

JuniorHsu replied:
and make the funciton `const`

Winbobob replied:
Thanks, just made the change accordingly!

JuniorHsu commented on source/extensions/filters/network/zookeeper_proxy/filter.h:
```diff
@@ -122,12 +188,26 @@ class ZooKeeperFilterConfig {
+  int32_t getOpCodeIndex(LatencyThreshold_Opcode opcode);
+  absl::flat_hash_map<int32_t, std::chrono::milliseconds>
+  parseLatencyThresholds(LatencyThresholdList& latency_thresholds);
+  std::chrono::milliseconds getDefaultLatencyThreshold(
```
no need for the parameter. just `getDefaultLatencyThreshold()`

Winbobob replied:
Ah, good point, thanks!

JuniorHsu commented on api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:
```diff
@@ -39,4 +40,60 @@ message ZooKeeperProxy {
+  //
+  // https://sre.google/workbook/implementing-slos/
+  //
+  // If no latency threshold is set, the fast/slow response counters will always be 0.
```
// The fast/slow response stats are emitted only if at least one latency threshold is set.

Winbobob replied:
I think the admin interface `/stats` will show all counters include ones with value 0.

JuniorHsu replied:
sure

JuniorHsu commented on test/extensions/filters/network/zookeeper_proxy/filter_test.cc:
```diff
@@ -30,7 +31,19 @@ class ZooKeeperFilterTest : public testing::Test {
+    threshold->mutable_threshold()->set_nanos(200000000); // 200 millisecond
+
+    Protobuf::RepeatedPtrField<
+        envoy::extensions::filters::network::zookeeper_proxy::v3::LatencyThreshold>&
```
I think we can use `LatencyThresholdList` here and there instead of full name

JuniorHsu replied:
I'm confused why we need a ref here. and tailing underscore is the naming style of data member.

Winbobob replied:
Sorry we do not need a ref here and pass the value to the func directly.

JuniorHsu commented on test/extensions/filters/network/zookeeper_proxy/filter_test.cc:
```diff
@@ -30,7 +31,19 @@ class ZooKeeperFilterTest : public testing::Test {
-    config_ = std::make_shared<ZooKeeperFilterConfig>(stat_prefix_, 1048576, scope_);
+    // Populate latency threshold for Multi opcode. The latency thresholds of all other opcode
+    // fallback to the default threshold. Since we did not specify the default threshold, it
+    // fallbacks to 100 millisecond.
```
It's implicit for testing the default behavior. could we have 2 new tests?

like
```
config =  std::make_shared<ZooKeeperFilterConfig>(stat_prefix_, 1048576, latency_thresholds_, scope_);
EXPECT_EQ(config.errorbudgetDecision(opcode, 99), FAST);
EXPECT_EQ(config.errorbudgetDecision(opcode, 101), SLOW);
```
one test have default value, another doesn't

Winbobob replied:
Thanks, just add more tests in both `config_test.cc` and `filter_test.cc` to test below four config scenario:
1. No latency thresholds specified
2. Only default latency threshold specified
3. Only non-default latency threshold specified
4. Both default and non-default latency thresholds specified

JuniorHsu reviewed (changes requested):
Thanks! I'd like to hold the r+ for the final round but the commit looks very good.

Winbobob commented:
Hi @adisuissa, would you mind checking the diff again, since @JuniorHsu just approved it? Thank you!

adisuissa commented on api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:
```diff
@@ -39,4 +40,60 @@ message ZooKeeperProxy {
+
+message LatencyThreshold {
+  enum Opcode {
+    Default = 0;
```
Looking at the [ZooDef.java](https://github.com/apache/zookeeper/blob/89c1831f84891f425f1fa9224210587124f1c1ec/zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java#L38) declaration, this seems to be reserved for `notification`. I think the "default" value should be explicitly defined in its own field, and these will serve as overrides for a specific Opcode.

Winbobob replied:
We cannot make the proto value 1:1 mapping to the opcode value since some opcode values are negative. So we maintains a map in the code.

If we move the default threshold to a separate field within the `LatencyThreshold` message, I think  it will only serve as the default value of one specific opcode instead of the default value of all opcode since we can repeat the `LatencyThreshold` message.

If we move the default threshold under the `ZooKeeperProxy` message, then there are two fields related to latency threshold, will it confuse people?

Winbobob replied:
cc @adisuissa, any thoughts? Thanks

Winbobob replied:
Yeah, agree. Just moved the default latency threshold under `ZooKeeperProxy`.

adisuissa commented on api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:
```diff
@@ -39,4 +40,60 @@ message ZooKeeperProxy {
+message LatencyThreshold {
+  enum Opcode {
+    Default = 0;
+    Connect = 1;
```
Not a ZooKeeper expert, but this doesn't appear to be defined in [ZooDefs.java](https://github.com/apache/zookeeper/blob/89c1831f84891f425f1fa9224210587124f1c1ec/zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java#L36), and some of the other opcodes' values are shifted by +1.
Is this deviation necessary, or can we keep the them in sync?

Winbobob replied:
Unfortunately, we cannot keep them in sync since there are [some negative opcode values](https://github.com/apache/zookeeper/blob/89c1831f84891f425f1fa9224210587124f1c1ec/zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java#L98-L102) that cannot happen in proto.
Specifically, I followed the opcode order [here](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/network/zookeeper_proxy/decoder.h#L28). There are still several opcodes in [ZooDefs.java](https://github.com/apache/zookeeper/blob/89c1831f84891f425f1fa9224210587124f1c1ec/zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java#L36) but not in the previous list, I will try to support them in my next PR.

Winbobob replied:
Yeah, the names are consistent between the two.

adisuissa commented on source/extensions/filters/network/zookeeper_proxy/filter.cc:
```diff
@@ -19,55 +21,169 @@ namespace NetworkFilters {
+  }
+
+  std::chrono::milliseconds default_latency_threshold(100);
+  auto it = latency_threshold_map_.find(-999);
```
This "magic" number should either be removed (maybe by using an explicit default value field), or it should be properly documented here.

Winbobob replied:
Yeah, I documented [here](https://github.com/envoyproxy/envoy/pull/26594/files#diff-a4f1b5b91a56b1929b870a98140d5ca6fe188a446c1acb64e42bc0fbdcbc5a94R131), which is a mapping between proto value and opcode value.

adisuissa reviewed (commented):
Thanks! Left a couple of comments.

phlax commented:
@Winbobob i think there is a genuine fail in the proto linting, but it seems the tool is not working as expected (#27147) - i will investigate this further

adisuissa commented on source/extensions/filters/network/zookeeper_proxy/filter.cc:
```diff
@@ -18,56 +20,152 @@ namespace Extensions {
+      {LatencyThresholdOverride::SetWatches, 101},
+      {LatencyThresholdOverride::GetEphemerals, 103},
+      {LatencyThresholdOverride::GetAllChildrenNumber, 104},
+      {LatencyThresholdOverride::SetWatches2, 105}};
```
I may have missed this, but can there be a test that validates what happens when the incoming value is not in the table (the config may be rejected, but just to make sure it is covered).

Winbobob replied:
Yeah, there is a [config test for undefined opcode](https://github.com/envoyproxy/envoy/pull/26594/files#diff-b680ac531c7b3eb41602f7d86e5d6bb8271920609161c52bf80eaa7898603913R46-R58).

Winbobob commented on source/extensions/filters/network/zookeeper_proxy/filter.h:
```diff
@@ -122,12 +188,24 @@ class ZooKeeperFilterConfig {
   const Stats::StatName unknown_opcode_latency_;
 
+  ErrorBudgetResponseType errorBudgetDecision(const OpCodes opcode,
+                                              const std::chrono::milliseconds latency) const;
```
Yeah, this func compares latency with threshold. The func supposes not to set the latency value.

Winbobob commented on test/extensions/filters/network/zookeeper_proxy/filter_test.cc:
```diff
@@ -30,7 +31,21 @@ class ZooKeeperFilterTest : public testing::Test {
 
   void initialize() {
-    config_ = std::make_shared<ZooKeeperFilterConfig>(stat_prefix_, 1048576, scope_);
+    std::chrono::milliseconds default_latency_threshold(100);
+    LatencyThresholdOverrideList latency_threshold_overrides;
+
+    initializeHelper(default_latency_threshold, latency_threshold_overrides);
+  }
+
+  void initialize(std::chrono::milliseconds default_latency_threshold,
+                  LatencyThresholdOverrideList& latency_threshold_overrides) {
+    initializeHelper(default_latency_threshold, latency_threshold_overrides);
+  }
+
+  void initializeHelper(std::chrono::milliseconds default_latency_threshold,
+                        LatencyThresholdOverrideList& latency_threshold_overrides) {
```
Since there are lots of `initialize()` func call already in the tests without parameters, I created the overload funcs. Otherwise, I need to change each `initialize()` func call.

adisuissa replied:
Sorry, I don't follow.
Can something similar to the following be used:
```
void initialize(std::chrono::milliseconds default_latency_threshold = 100,
                       const LatencyThresholdOverrideList& latency_threshold_overrides = {}) {
    ....
}
```
(maybe needs a bit of massaging of the values after the `=`)

Winbobob replied:
Sorry, you are right, was thinking about Golang 🤦

Winbobob commented:
Hi @adisuissa, just addressed all your comments, please review it again, thanks!

adisuissa reviewed (commented):
Thanks!
One small comment on the API. Other than that LGTM!
/assign-from @envoyproxy/senior-maintainers

alyssawilk commented:
looks like this is extension code and mostly stats, so throwing it over to Josh to review and merge

Winbobob commented:
cc @jmarantz, could you review the code? Thank you!

jmarantz commented on source/extensions/filters/network/zookeeper_proxy/decoder.cc:
```diff
@@ -617,7 +617,7 @@ void DecoderImpl::decode(Buffer::Instance& data, DecodeType dtype, uint64_t full
 
 void DecoderImpl::parseConnectResponse(Buffer::Instance& data, uint64_t& offset, uint32_t len,
-                                       const std::chrono::milliseconds& latency) {
+                                       const std::chrono::milliseconds latency) {
```
why all these changes (const ref --> const)?

Is that related to the functional changes you are making? Maybe you could peel those off to a separate PR?

JuniorHsu replied:
copying small object is preferred for perf at least most of the repo is in favors of copy chrono type. Hence, I asked to remove ref while he introduce new function with ref to `std::chrono::milliseconds`. Yeah small diff is preferred but I thought consistency might be another thing to consider.

jmarantz replied:
consistency is fine but why don't break that out into a separate PR. It's not related to this one.

JuniorHsu replied:
this PR have new functions with the chrono types so splitting forces us to either
(a) bring more `const std::chrono::milliseconds&` as , or  
(b) make inconsistency inside zk filter to have both ref and copying

Another PR is fine I agree so I won't insist

Winbobob replied:
Thank you for the discussion!
I will revert unrelated changes and create another PR after this one.

jmarantz reviewed (commented):
I left a few minor comments; I feel like I'm going to need to get to a big monitor to review this and I don't have one available right now.

jmarantz commented on source/extensions/filters/network/zookeeper_proxy/filter.h:
```diff
@@ -106,7 +171,9 @@ class ZooKeeperFilterConfig {
   // instance.
   struct OpCodeInfo {
-    Stats::Counter* counter_;
+    Stats::Counter* resp_counter_;
```
ok, sounds good.

Winbobob replied:
Hi @jmarantz, I have added a comment in the code with more details. Please check it again, thank you!

jmarantz replied:
thanks this helps.

Winbobob commented on source/extensions/filters/network/zookeeper_proxy/filter.h:
```diff
@@ -81,7 +83,61 @@ namespace ZooKeeperProxy {
   COUNTER(removewatches_resp)                                                                      \
   COUNTER(check_resp)                                                                              \
-  COUNTER(watch_event)
+  COUNTER(watch_event)                                                                             \
```
Unfortunately, in order to calculate the error budget for each opcode, we have to introduce these counters.
However, they should be included in only one scope, I think.

Winbobob replied:
Hi @jmarantz, not sure whether the above comment makes sense to you.

Btw, I have addressed all other comments. Please review again whenever you have a chance, thank you!

Winbobob commented:
/retest

Winbobob commented:
> I left a few more comments; I feel like I'm going to need to get to a big monitor to finish reviewing this and I don't have one available right now.
> 
> 

Hi @jmarantz, not sure whether you have a chance to check the code again. Thank you!

jmarantz commented:
Yes, no need to keep pinging me I will take a look. We do get auto-reminded of what's pending so this will not fall off my radar but I sometimes have other time-critical things that need to be done.

jmarantz commented on test/extensions/filters/network/zookeeper_proxy/config_test.cc:
```diff
@@ -43,6 +43,91 @@ max_packet_bytes: -1
+  )EOF";
+
+  ZooKeeperProxyProtoConfig proto_config;
+  EXPECT_THROW(TestUtility::loadFromYamlAndValidate(yaml, proto_config), EnvoyException);
```
Scan for text identifying the error is included in the exception using EXPECT_THROW_WITH_MESSAGE?

jmarantz commented on test/extensions/filters/network/zookeeper_proxy/config_test.cc:
```diff
@@ -60,6 +145,139 @@ stat_prefix: test_prefix
+  TestUtility::loadFromYamlAndValidate(yaml, proto_config);
+
+  testing::NiceMock<Server::Configuration::MockFactoryContext> context;
+  ZooKeeperConfigFactory factory;
```
I see a pattern here; factor out the common context here into a test class and use TEST_F?

jmarantz commented on test/extensions/filters/network/zookeeper_proxy/config_test.cc:
```diff
@@ -60,6 +145,139 @@ stat_prefix: test_prefix
+  Network::FilterFactoryCb cb = factory.createFilterFactoryFromProto(proto_config, context);
+  Network::MockConnection connection;
+  EXPECT_CALL(connection, addFilter(_));
+  cb(connection);
```
should we be checking that the latency threshold made it to the actual config object?

jmarantz commented on test/extensions/filters/network/zookeeper_proxy/config_test.cc:
```diff
@@ -60,6 +145,139 @@ stat_prefix: test_prefix
+  - opcode: SetWatches2
+    threshold:
+      nanos: 176000000
+  )EOF";
```
can we refactor this test so we can check that these configs actually make it through?

I think you can do this deeper test much more tersely than even what you have here, which IIUC is only verifying that you don't throw.

You can hake a test-fixture helper method to formulate a coherent config fragment with just one setting, and then verify that setting made it through correctly.

jmarantz reviewed (commented):
mostly looking good; just a few nits/testing concerns.

jmarantz commented on test/extensions/filters/network/zookeeper_proxy/filter_test.cc:
```diff
@@ -29,8 +30,12 @@ class ZooKeeperFilterTest : public testing::Test {
 public:
   ZooKeeperFilterTest() { ENVOY_LOG_MISC(info, "test"); }
```
pre-existing, but can we drop this message?

jmarantz reviewed (commented):
High level question: this is a ton of new stats.

Do you want to including a configuration knob about whether to include them? otherwise every Zookeeper will be carrying these and maintaining them, and they are not free.

Winbobob commented:
> High level question: this is a ton of new stats.
> 
> Do you want to including a configuration knob about whether to include them? otherwise every Zookeeper will be carrying these and maintaining them, and they are not free.

Thank you for the suggestions! Just added a new boolean `enable_latency_threshold_metrics` in the proto to control the metrics emitting.

jmarantz commented on api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:
```diff
@@ -42,19 +42,23 @@ message ZooKeeperProxy {
 
+  // Whether to emit latency threshold metrics. If not set, defaults to false.
+  // If false, setting `default_latency_threshold` and `latency_threshold_overrides` will not have effect.
+  bool enable_latency_threshold_metrics = 4;
```
I think we'd want to leave the previous numeric IDs the same and make this new one be number 6.

Winbobob replied:
Since both 5 and 6 are also newly added, I think it should be ok to make `enable_latency_threshold_metrics` as 4?
The main reason is that with current order, the proto doc could be a bit easier to read. We put the toggle (`enable_latency_threshold_metrics`) first, then the detailed settings.

jmarantz replied:
Oh sorry I was looking only at the diff in the most recent commit. What you have is fine :)

jmarantz commented on docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst:
```diff
@@ -31,8 +31,9 @@ in the configuration snippet below:
-Every configured ZooKeeper proxy filter has statistics rooted at *<stat_prefix>.zookeeper.*. The
-following counters are available:
+Every configured ZooKeeper proxy filter has statistics rooted at *<stat_prefix>.zookeeper.*.
+*_resp_fast* and *_resp_slow* metrics will only be emitted when enabling the latency threshold metrics.
```
slight re-wording suggestion: 

will only be emitted when ``enable_latency_threshold_metrics`` is set to `true`.

jmarantz commented on source/extensions/filters/network/zookeeper_proxy/filter.cc:
```diff
@@ -123,8 +128,35 @@ void ZooKeeperFilterConfig::initOpCode(OpCodes opcode, Stats::Counter& resp_coun
 int32_t ZooKeeperFilterConfig::getOpCodeIndex(LatencyThresholdOverride_Opcode opcode) {
-  auto it = opcodeMap().find(opcode);
-  if (it != opcodeMap().end()) {
+  static OpcodeMap opcode_map = {{LatencyThresholdOverride::Connect, 0},
```
this pattern causes a non-deterministic destructor order fiasco. The previous impl you had was closer to what I was thinking. Just need a temp for the map:
```
const OpcodeMap& opcode_map = opcodeMap();
auto it = opcode_map.find(opcode);
if (it != opcode_map.end()) {
...
```

Winbobob replied:
Sorry, I misunderstood your previous comment, just updated the code.

jmarantz commented on source/extensions/filters/network/zookeeper_proxy/filter.h:
```diff
@@ -146,7 +146,7 @@ struct ZooKeeperProxyStats {
 };
 
-enum class ErrorBudgetResponseType { FAST, SLOW };
+enum class ErrorBudgetResponseType { FAST, SLOW, NONE };
```
I missed this before, but let's spell the enum values as `Fast`, `Slow`, `None`.

The style guide allows all-caps but that's problematic because that collides with the convention for `#define` and those are not scoped, so if you get a `#define NONE 0` pulled into your build then your code stops compiling. As this enum is new to this PR let's spell them without all-caps.

jmarantz commented on test/extensions/filters/network/zookeeper_proxy/config_test.cc:
```diff
@@ -120,162 +173,101 @@ max_packet_bytes: 1048576
+  TestUtility::loadFromYamlAndValidate(yaml, proto_config_);
+
+  EXPECT_EQ(proto_config_.stat_prefix(), "test_prefix");
+  EXPECT_EQ(proto_config_.max_packet_bytes().value(), 1048576);
```
it's hard to tell; are we now covering the setting of all the fields?

Winbobob replied:
Yes, we cover all configs in this test case, including
- stat_prefix
- max_packet_bytes
- enable_latency_threshold_metrics
- default_latency_threshold
- 26 latency_threshold_overrides

jmarantz reviewed (approved):
Thanks for following up on all these issues!

Winbobob commented:
Thank you so much for reviewing the code and providing valuable feedback!
Learnt a lot!
